src/parser.rs

```diff
@@ -0,0 +1,978 @@
+                Token::RParent => {
+                    self.go_next_token();
+                    break;
+                }
+                _ => {
+                    let expr = self.expr()?;
+                    args.push(expr);
+                }
+            }
+        }
+
+        self.expect(&Token::Semi)?;
+        Ok(Node::Call(name, Box::new(args)))
+    }
+
+    fn parse_assign(&mut self) -> ParseResult<Node> {
+        let id_name: String = self.consume_id()?;
+        self.expect(&Token::Assign)?;
+        let expr: Node = self.expr()?;
+        self.expect(&Token::Semi)?;
+
+        Ok(Node::Assign(id_name, Box::new(expr)))
+    }
+
+    fn expr(&mut self) -> ParseResult<Node> {
+        self.equal()
+    }
+
+    fn equal(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.relation()?;
+        self.check_invalid(&lhs)?;
+
+        loop {
+            if !self.check_vec(vec![Token::Eq, Token::Ne]) {
+                break;
+            }
+            let op: Token = self.get_token();
+            self.go_next_token();
+
+            if let &Token::Eq = &op {
+                lhs = Node::Eq(Box::new(lhs), Box::new(self.relation()?));
+            } else if let &Token::Ne = &op {
+                lhs = Node::Ne(Box::new(lhs), Box::new(self.relation()?));
+            }
+        }
+
+        Ok(lhs)
+    }
+
+    fn relation(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.shift()?;
+        self.check_invalid(&lhs)?;
+
+        loop {
+            if !self.check_vec(vec![Token::Lt, Token::Gt, Token::Le, Token::Ge]) {
+                break;
+            }
+            let op: Token = self.get_token();
+            self.go_next_token();
+
+            if let &Token::Lt = &op {
+                lhs = Node::Lt(Box::new(lhs), Box::new(self.relation()?));
+            } else if let &Token::Gt = &op {
+                lhs = Node::Gt(Box::new(lhs), Box::new(self.relation()?));
+            } else if let &Token::Le = &op {
+                lhs = Node::Le(Box::new(lhs), Box::new(self.relation()?));
+            } else if let &Token::Ge = &op {
+                lhs = Node::Ge(Box::new(lhs), Box::new(self.relation()?));
+            }
+        }
+
+        Ok(lhs)
+    }
+
+    fn shift(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.add_sub()?;
+        self.check_invalid(&lhs)?;
+
+        loop {
+            if self.check(&Token::Shl) {
+                self.go_next_token();
+                lhs = Node::Shl(Box::new(lhs), Box::new(self.add_sub()?));
+            } else if self.check(&Token::Shr) {
+                self.go_next_token();
+                lhs = Node::Shr(Box::new(lhs), Box::new(self.add_sub()?));
+            } else {
+                break;
+            }
+        }
+
+        Ok(lhs)
+    }
+
+    fn add_sub(&mut self) -> ParseResult<Node> {
+        let mut lhs: Node = self.mul_div()?;
+        self.check_invalid(&lhs)?;
```

Comment:
I expect, that after calling `check_invalid` `lhs` on the next line will be invalid. Maybe, `ensure_valid` will solve the ambiguity.